dialogue-core/src/main/java/com/palantir/dialogue/core/NodeSelectionStrategyChannel.java

```diff
@@ -101,21 +134,68 @@ private static LimitedChannel getUpdatedNodeSelectionStrategy(
                     PinUntilErrorNodeSelectionStrategyChannel previousPinUntilError =
                             (PinUntilErrorNodeSelectionStrategyChannel) previousNodeSelectionStrategy;
-                    return PinUntilErrorNodeSelectionStrategyChannel.of(
-                            Optional.of(previousPinUntilError.getCurrentChannel()),
-                            updatedStrategy,
-                            channels,
-                            pinuntilerrorMetrics,
-                            random,
-                            channelName);
+                    return channelBuilder
+                            .channel(PinUntilErrorNodeSelectionStrategyChannel.of(
+                                    Optional.of(previousPinUntilError.getCurrentChannel()),
+                                    updatedStrategy,
+                                    channels,
+                                    pinuntilerrorMetrics,
+                                    random,
+                                    channelName))
+                            .build();
                 }
-                return PinUntilErrorNodeSelectionStrategyChannel.of(
-                        Optional.empty(), updatedStrategy, channels, pinuntilerrorMetrics, random, channelName);
-            case ROUND_ROBIN:
+                return channelBuilder
+                        .channel(PinUntilErrorNodeSelectionStrategyChannel.of(
+                                Optional.empty(), updatedStrategy, channels, pinuntilerrorMetrics, random, channelName))
+                        .build();
+            case BALANCED:
                 // When people ask for 'ROUND_ROBIN', they usually just want something to load balance better.
                 // We used to have a naive RoundRobinChannel, then tried RandomSelection and now use this heuristic:
-                return new BalancedNodeSelectionStrategyChannel(channels, random, tick, metrics, channelName);
+                return channelBuilder
+                        .channel(new BalancedNodeSelectionStrategyChannel(channels, random, tick, metrics, channelName))
+                        .build();
+            case UNKNOWN:
         }
         throw new SafeRuntimeException("Unknown NodeSelectionStrategy", SafeArg.of("unknown", updatedStrategy));
     }
+
+    @VisibleForTesting
+    static Optional<DialogueNodeSelectionStrategy> getFirstKnownStrategy(
+            List<DialogueNodeSelectionStrategy> strategies) {
+        for (DialogueNodeSelectionStrategy strategy : strategies) {
+            if (!strategy.equals(DialogueNodeSelectionStrategy.UNKNOWN)) {
+                return Optional.of(strategy);
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Value.Immutable
+    interface NodeSelectionChannel {
+        DialogueNodeSelectionStrategy strategy();
+
+        LimitedChannel channel();
+
+        ImmutableList<LimitedChannel> hostChannels();
```

Comment:
why are hostChannels saved here rather than just using the instance field?

Reply:
Ya we need to keep track of the host channels so that we can recreate the underlying channel in response to a server's node-selection-strategy response